draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+
+The queue must be thread-safe and coroutine-safe.
+
+Services
+++++++++
+
+Services are stateful, long running tasks which are used by Celery to perform
+its internal operations.
+
+Some services publish :term:`messages <Message>` to brokers,
+others consume :term:`messages <Message>` from them.
+Other services are used to calculate optimal scheduling of tasks, routing,
+logging and even executing tasks.
+
+Users may create their own services as well.
+
+Internal Services
++++++++++++++++++
+
+The Worker defines internal services to ensure it's operation and to provide
+support for it's features.
+
+The exact API for each service will be determined in another CEP.
+
+This list of internal services is not final.
+Other internal services may be defined in other CEPs.
+
+Task Execution
+~~~~~~~~~~~~~~
+
+The ``Task Execution`` service is responsible for executing all Celery
+:ref:`tasks <draft/celery-5-high-level-architecture:Tasks>`.
+
+It consumes tasks from the
+:ref:`draft/celery-5-high-level-architecture:Internal Tasks Queue`,
+executes them and enqueues the results into the
+:ref:`draft/celery-5-high-level-architecture:Internal Results Queue`.
+
+The service supervises how many tasks are run concurrently and limits the
+number of concurrent tasks to the configured amount in accordance to the
+:ref:`draft/celery-5-high-level-architecture:Concurrency Budget`.
+
+The service also attempts to saturate all of the available resources by
+scheduling as many as :ref:`draft/celery-5-high-level-architecture:I/O Bound Tasks`
+and :ref:`draft/celery-5-high-level-architecture:CPU Bound Tasks` as possible.
+
+
+Consumer
+~~~~~~~~
+
+The ``Consumer`` service consumes :term:`messages <Message>` from one or many
+:ref:`Routers <draft/celery-5-high-level-architecture:Router>` or
+:term:`Message Brokers <Message Broker>`.
+
+The service enqueues the consumed :term:`messages <Message>`
+into the appropriate :ref:`draft/celery-5-high-level-architecture:Internal Tasks Queue`
+according to the task's type.
+
+Result Publisher
+~~~~~~~~~~~~~~~~
+
+The ``Result Publisher`` service consumes results from the
+:ref:`draft/celery-5-high-level-architecture:Internal Results Queue` and
+publishes them to the :ref:`draft/celery-5-high-level-architecture:Router`'s
+:ref:`draft/celery-5-high-level-architecture:Inbox Queue`.
+
+Maximal Concurrency Budget
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``Maximal Concurrency Budget`` service runs the user's concurrency budget
+strategies and notifies the :ref:`tasks <draft/celery-5-high-level-architecture:Task Execution>`
+service of changes in concurrency.
+
+Tasks
++++++
+
+Tasks are short running, have a defined purpose and are triggered in response
+to messages.
+
+Celery declares some tasks for internal usage.
+
+Users will create their own tasks for their own use.
+
+Deduplication
+~~~~~~~~~~~~~
+
+Some Tasks are not idempotent and may not run more than once.
+
+Users may define a deduplication policy to help Celery discard duplicated
```

Comment:
So we should provide an API for that, in the same manner that retries are handled?